enterprise/internal/database/perms_store.go

```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
 	}
 	return m
 }
+
+// ListUserPermissions gets the list of accessible repos for the provided user, along with the reason
+// and timestamp for each permission.
+func (s *permsStore) ListUserPermissions(ctx context.Context, userID int32, args *ListUserPermissionsArgs) (perms []*UserPermission, err error) {
+	ctx, save := s.observe(ctx, "ListUserPermissions", "")
+	defer save(&err)
+
+	// Set actor with provided userID to context.
+	ctx = actor.WithActor(ctx, actor.FromUser(userID))
+
+	db := database.NewDBWith(s.logger, s.Store)
+
+	// NOTE(naman): `List` uses the actor from ctx to apply authz conds
+	// to fetch list of repos which are visible to the actor.
+	repos, err := db.Repos().List(ctx, database.ReposListOptions{
+		PaginationArgs: args.PaginationArgs,
+		Query:          args.Query,
+	})
+
+	authzParams, err := database.GetAuthzQueryParameters(ctx, db)
+	if err != nil {
+		return nil, err
+	}
+
+	if authzParams.BypassAuthz {
+		// NOTE(naman): If Authz conds were bypassed then no need to fetch
+		// perm details for the repos. Show permission reason as "Site Admin"
+		// or "Authentication Bypassed" based on query params used.
+
+		reason := UserRepoPermissionReasonAuthzBypass
+
+		if authzParams.BypassAuthzReason == database.BypassAuthzReasonSiteAdmin {
+			reason = UserRepoPermissionReasonSiteAdmin
+		}
+
+		for _, repo := range repos {
+			perms = append(perms, &UserPermission{Repo: repo, Reason: reason})
+		}
+	} else {
+		// NOTE(naman): If Authz conds were not bypassed then find perm details
+		// for each of the repos and show correct reason and updatedAt.
+
+		repoIDs := []int32{}
```

Comment:
We know the size already. Initializing upfront will save us a few memory allocations:
```suggestion
		repoIDs := make([]int32, 0, len(repos))
```